### ISSUE TITLE
eddyqc_text vs eddyqc_all

### ISSUE THREAD

**Issue by araikes** [BIDS-Apps/MRtrix3_connectome#71]
Not sure if this is a bug related to this script or _MRtrix3_. 

I swapped eddyqc_text for eddyqc_all, expecting to get `.nii.gz` files along with the extension-less files. However, I get the same files (no `.nii.gz`) regardless of whether it's `--eddyqc_text` or `eddyqc_all`

Relatedly, since `eddy_quad` is being run somewhere in the process, should there be a `.pdf` and `.json` file with the single-subject report and quality metrics (respectively)? Summary motion parameters would be beneficial to extract (a la fMRIPrep) for regression.

https://github.com/BIDS-Apps/MRtrix3_connectome/blob/3c712dfb365bf7c7703620e11d34fe1f4e48261b/mrtrix3_connectome.py#L820

**Comment by Lestropie** [comment]
> However, I get the same files (no `.nii.gz`) regardless of whether it's `-eddyqc_text` or `-eddyqc_all`

Extra files that should be grabbed when `-eddyqc_all` is used are [here](https://github.com/MRtrix3/mrtrix3/blob/3.0_RC3/bin/dwipreproc#L129). Only things that come to mind would be either the names of those files changing between FSL versions, or being written with `.nii` rather than `.nii.gz` extension (some FSL commands honour `FSLOUTPUTTYPE` while others ignore it, and this could conceivably change between FSL versions also). Running with `--debug` (or `-nocleanup` outside the container) might give some insight into why these are being missed.

> Relatedly, since eddy_quad is being run somewhere in the process, should there be a .pdf and .json file with the single-subject report and quality metrics (respectively)?

I only have `eddy_quad` being invoked in `dwipreproc` within the `dev` branch of *MRtrix3* (now `dwifslpreproc`; code [here](https://github.com/MRtrix3/mrtrix3/blob/8269f6e9304f2dd2a09bc0a21736ced51c0a0ced/bin/dwifslpreproc#L856-L886) and [here](https://github.com/MRtrix3/mrtrix3/blob/8269f6e9304f2dd2a09bc0a21736ced51c0a0ced/bin/dwifslpreproc#L1088-L1092)); this change has not yet made it into *MRtrix3* `master` or this tool.

**Comment by araikes** [comment]
> Running with `--debug` (or `-nocleanup` outside the container) might give some insight into why these are being missed.
> 

See below

> I only have `eddy_quad` being invoked in `dwipreproc` within the `dev` branch of _MRtrix3_ (now `dwifslpreproc`; code [here](https://github.com/MRtrix3/mrtrix3/blob/8269f6e9304f2dd2a09bc0a21736ced51c0a0ced/bin/dwifslpreproc#L856-L886) and [here](https://github.com/MRtrix3/mrtrix3/blob/8269f6e9304f2dd2a09bc0a21736ced51c0a0ced/bin/dwifslpreproc#L1088-L1092)); this change has not yet made it into _MRtrix3_ `master` or this tool.

Fair enough. I thought that `eddy_quad` was part of how the `eddyqc/` objects were generated.

Eddy commands that were run using the script:
```
mrtrix3_connectome.py: [DEBUG] run.command() (from mrtrix3_connectome.py:820): To execute: [['/usr/bin/env', 'python', '/mrtrix3/bin/dwipreproc', 'dwipreproc_in.mif', 'dwi_preprocessed.mif', '-se_epi', 'se_epi.mif', '-align_seepi', '-eddy_options', ' --repol --estimate_move_by_susceptibility', '-rpe_header', '-eddyqc_all', 'eddyqc/', '-info']]
Command:  dwipreproc dwipreproc_in.mif dwi_preprocessed.mif  -se_epi se_epi.mif -align_seepi -eddy_options " --repol --estimate_move_by_susceptibility" -rpe_header -eddyqc_all eddyqc/
          dwipreproc:
          dwipreproc: Note that this script makes use of commands / algorithms that have relevant articles for citation; INCLUDING FROM EXTERNAL SOFTWARE PACKAGES. Please consult the help page (-help option) for more information.
          dwipreproc:
          dwipreproc: Loading header for image file '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc_in.mif'
          dwipreproc: Generated temporary directory: /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc-tmp-KT8G0C/
          Command:  mrconvert /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc_in.mif /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc-tmp-KT8G0C/dwi.mif
                    mrconvert: [100%] copying from "/outputs/m...mp-Z13G0W/dwipreproc_in.mif" to "/outputs/m...ipreproc-tmp-KT8G0C/dwi.mif"
          dwipreproc: Loading header for image file '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/se_epi.mif'
          Command:  mrconvert /outputs/mrtrix3_connectome.py-tmp-Z13G0W/se_epi.mif /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc-tmp-KT8G0C/se_epi.mif
                    mrconvert: [100%] copying from "/outputs/m...me.py-tmp-Z13G0W/se_epi.mif" to "/outputs/m...eproc-tmp-KT8G0C/se_epi.mif"
          dwipreproc: Changing to temporary directory (/outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc-tmp-KT8G0C/)
          dwipreproc: Loading header for image file 'dwi.mif'
          dwipreproc: Loading header for image file 'se_epi.mif'
          dwipreproc: Command: '/mrtrix3/bin/mrinfo dwi.mif -shell_bvalues' (piping data to local storage)
          dwipreproc: Result: 0 1000
          Command:  dirstat dwi.mif -output asym
          dwipreproc: No phase-encoding contrast present in SE-EPI images; will examine again after combining with DWI b=0 images
          Command:  dwiextract dwi.mif - -bzero | mrcat - se_epi.mif se_epi_dwibzeros.mif -axis 3
                    dwiextract: [100%] extracting volumes
                    mrcat: [100%] concatenating "/tmp/mrtrix-tmp-mRrcYb.mif"
                    mrcat: [100%] concatenating "se_epi.mif"
                    dwipreproc: Loading header for image file 'se_epi_dwibzeros.mif'
          Command:  mrinfo dwi.mif -export_grad_mrtrix grad.b
          Command:  mrconvert se_epi_dwibzeros.mif topup_in.nii -strides -1,+2,+3,+4 -export_pe_table topup_datain.txt
                    mrconvert: [100%] copying from "se_epi_dwibzeros.mif" to "topup_in.nii"
          Command:  topup --imain=topup_in.nii --datain=topup_datain.txt --out=field --fout=field_map.nii --config=/opt/fsl/etc/flirtsch/b02b0.cnf
          Command:  mrinfo dwi.mif -export_pe_eddy applytopup_config.txt applytopup_indices.txt
          Command:  dwiextract dwi.mif -pe 0,-1,0,0.0597 - | mrconvert - dwi_pe_1.nii -json_export dwi_pe_1.json
                    dwiextract: [100%] extracting volumes
                    mrconvert: [100%] copying from "/tmp/mrtrix-tmp-u5nMaI.mif" to "dwi_pe_1.nii"
                    Command:  applytopup --imain=dwi_pe_1.nii --datain=applytopup_config.txt --inindex=1 --topup=field --out=dwi_pe_1_applytopup.nii --method=jac
          Command:  mrconvert dwi_pe_1_applytopup.nii dwi_pe_1_applytopup.mif -json_import dwi_pe_1.json
                    mrconvert: [100%] copying from "dwi_pe_1_applytopup.nii" to "dwi_pe_1_applytopup.mif"
          Command:  dwi2mask dwi_pe_1_applytopup.mif - | maskfilter - dilate - | mrconvert - eddy_mask.nii -datatype float32 -strides -1,+2,+3
                    dwi2mask: [100%] preloading data for "dwi_pe_1_applytopup.mif"
                    dwi2mask: [100%] finding min/max of "mean b=0 image"
                    dwi2mask: [done] optimising threshold
                    dwi2mask: [100%] thresholding
                    dwi2mask: [100%] finding min/max of "mean b=1000 image"
                    dwi2mask: [done] optimising threshold
                    dwi2mask: [100%] thresholding
                    dwi2mask: [done] computing dwi brain mask
                    dwi2mask: [done] applying mask cleaning filter
                    maskfilter: [100%] applying dilate filter to image -
                    mrconvert: [100%] copying from "/tmp/mrtrix-tmp-WUmKaP.mif" to "eddy_mask.nii"
                    Command:  mrconvert dwi.mif eddy_in.nii -strides -1,+2,+3,+4 -export_grad_fsl bvecs bvals -export_pe_eddy eddy_config.txt eddy_indices.txt
                    mrconvert: [100%] copying from "dwi.mif" to "eddy_in.nii"
          Command:  eddy_openmp --imain=eddy_in.nii --mask=eddy_mask.nii --acqp=eddy_config.txt --index=eddy_indices.txt --bvecs=bvecs --bvals=bvals --topup=field --repol --estimate_move_by_susceptibility --out=dwi_post_eddy
          Command:  mrconvert dwi_post_eddy.nii result.mif -strides -2,3,4,1 -fslgrad dwi_post_eddy.eddy_rotated_bvecs bvals
                    mrconvert: [100%] copying from "dwi_post_eddy.nii" to "result.mif"
          Function: shutil.copy('dwi_post_eddy.eddy_parameters', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_parameters')
          Function: shutil.copy('dwi_post_eddy.eddy_movement_rms', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_movement_rms')
          Function: shutil.copy('dwi_post_eddy.eddy_restricted_movement_rms', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_restricted_movement_rms')
          Function: shutil.copy('dwi_post_eddy.eddy_post_eddy_shell_alignment_parameters', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_post_eddy_shell_alignment_parameters')
          Function: shutil.copy('dwi_post_eddy.eddy_post_eddy_shell_PE_translation_parameters', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_post_eddy_shell_PE_translation_parameters')
          Function: shutil.copy('dwi_post_eddy.eddy_outlier_report', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_outlier_report')
          Function: shutil.copy('dwi_post_eddy.eddy_outlier_map', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_outlier_map')
          Function: shutil.copy('dwi_post_eddy.eddy_outlier_n_stdev_map', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_outlier_n_stdev_map')
          Function: shutil.copy('dwi_post_eddy.eddy_outlier_n_sqr_stdev_map', '/outputs/mrtrix3_connectome.py-tmp-Z13G0W/eddyqc/eddy_outlier_n_sqr_stdev_map')
          Command:  mrconvert result.mif /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwi_preprocessed.mif
                    mrconvert: [100%] copying from "result.mif" to "/outputs/m...Z13G0W/dwi_preprocessed.mif"
          dwipreproc: Changing back to original directory (/outputs/mrtrix3_connectome.py-tmp-Z13G0W)
          dwipreproc: Deleting temporary directory /outputs/mrtrix3_connectome.py-tmp-Z13G0W/dwipreproc-tmp-KT8G0C/
```

**Comment by Lestropie** [comment]
> I thought that eddy_quad was part of how the eddyqc/ objects were generated.

Nope; they (currently) just grab files generated by `eddy`. The "`-eddyqc_text`" and "`-eddyqc_all`" options existed and were named as such before I knew of EddyQC's existence :man_shrugging: 

> Eddy commands that were run using the script:

Unfortunately doesn't quite give me what I was hoping for; what I wanted was to be able to inspect the contents of the `dwipreproc` scratch directory, but it's getting erased because the MRtrix3_connectome script isn't propagating the `-nocleanup` option down to `dwipreproc`. Probably what you would need to do is go into the container with `-entrypoint=/bin/bash` and manually run `dwipreproc` on something, including the `-nocleanup` option. The contents of the scratch directory will hopefully tell us why it's failing to find those images.

I do however suspect it's a file extension issue, based on this line:
`          Command:  mrconvert dwi_post_eddy.nii result.mif -strides -2,3,4,1 -fslgrad dwi_post_eddy.eddy_rotated_bvecs bvals`
I don't think I've seen `eddy` honouring `FSLOUTPUTTYPE` and exporting `.nii` rather than `.nii.gz` before, which would certainly make `-eddyqc_all` not work given the current implementation *assumes* the output images will be `.nii.gz`. That's another change I'll have to make at the *MRtrix3* level.

**Comment by Lestropie** [comment]
Correction. It looks as though with the most up-to-date version of FSL, `eddy` no longer generates two of those three images automatically. There are instead explicit options `--cnr_maps` and `--residuals` for exporting the relevant image data. The outlier free image data is still automatically generated if `--repol` is specified.

Unfortunately every time interfaces of such tools are changed, it makes the job of interfacing with them a little harder, as all versions of the external software need to be supported...

**Comment by Lestropie** [comment]
Should be fixed by #62.
